llama.cpp

```diff
@@ -3909,9 +3909,57 @@ static bool llm_load_tensors(
     return true;
 }
 
+// check if the URL is a HuggingFace model, and if so, try to download it
+static void hf_try_download_model(std::string & url) {
+    bool is_url = false;
+
+    if (url.size() > 22) {
+        is_url = (url.compare(0, 22, "https://huggingface.co") == 0);
+    }
+
+    if (!is_url) {
+        return;
+    }
+
+    // Examples:
+    //
+    // https://huggingface.co/TheBloke/Mixtral-8x7B-Instruct-v0.1-GGUF/resolve/main/mixtral-8x7b-instruct-v0.1.Q2_K.gguf
+
+    std::string basename;
+    basename = url.substr(url.find_last_of("/\\") + 1);
+
+    LLAMA_LOG_INFO("%s: detected URL, attempting to download %s\n", __func__, basename.c_str());
+
+    {
+        const std::string cmd = "wget -q --show-progress -c -O " + basename + " " + url;
+        LLAMA_LOG_INFO("%s: %s\n", __func__, cmd.c_str());
+
+        const int ret = system(cmd.c_str());
+        if (ret == 0) {
+            url = basename;
+            return;
+        }
+    }
+
+    {
+        const std::string cmd = "curl -C - -f -o " + basename + " -L " + url;
+        LLAMA_LOG_INFO("%s: %s\n", __func__, cmd.c_str());
+
+        const int ret = system(cmd.c_str());
```

Comment:
Yeah, system and popen should not be used with arbitrary user input. Better to use fork/execlp or posix_spawn (Unix) and _spawnlp or CreateProcess (Windows). This is hard to do safely and portably in C or C++ without third-party dependencies such as glib.

Reply:
Yes, I dropped the idea. Will probably add a bash script that can be used something like this:

```bash
./main -m $(./examples/hf.sh --model-name TheBloke/Mixtral-8x7B-v0.1-GGUF --quant Q4_K_M) ...
```

and it will do all the `curl` and `wget` calls